// mat-video/esm5/app/video/directives/mat-video-source.directive.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,missingReturn,uselessCode} checked by tsc
 */
import { Directive, ElementRef, Host, Input } from '@angular/core';
import { MatVideoComponent } from '../video.component';
var MatVideoSourceDirective = /** @class */ (function () {
    function MatVideoSourceDirective(matVideoComponent, el) {
        this.matVideoComponent = matVideoComponent;
        this.el = el;
        this.src = null;
        this.type = null;
        this.init = true;
        this.video = matVideoComponent;
        this.source = el.nativeElement;
        this.init = false;
    }
    /**
     * @param {?} changes
     * @return {?}
     */
    MatVideoSourceDirective.prototype.ngOnChanges = /**
     * @param {?} changes
     * @return {?}
     */
    function (changes) {
        this.source.src = this.src;
        this.source.type = this.type;
        if (!this.init)
            this.video.load();
    };
    MatVideoSourceDirective.decorators = [
        { type: Directive, args: [{
                    selector: '[matVideoSource]'
                },] }
    ];
    /** @nocollapse */
    MatVideoSourceDirective.ctorParameters = function () { return [
        { type: MatVideoComponent, decorators: [{ type: Host }] },
        { type: ElementRef }
    ]; };
    MatVideoSourceDirective.propDecorators = {
        src: [{ type: Input }],
        type: [{ type: Input }]
    };
    return MatVideoSourceDirective;
}());
export { MatVideoSourceDirective };
if (false) {
    /** @type {?} */
    MatVideoSourceDirective.prototype.src;
    /** @type {?} */
    MatVideoSourceDirective.prototype.type;
    /** @type {?} */
    MatVideoSourceDirective.prototype.init;
    /** @type {?} */
    MatVideoSourceDirective.prototype.video;
    /** @type {?} */
    MatVideoSourceDirective.prototype.source;
    /** @type {?} */
    MatVideoSourceDirective.prototype.matVideoComponent;
    /** @type {?} */
    MatVideoSourceDirective.prototype.el;
}
